components/ai_chat/core/browser/tab_tracker_service_unittest.cc

```diff
@@ -0,0 +1,232 @@
+
+#include "base/task/current_thread.h"
+#include "base/test/task_environment.h"
+#include "brave/components/ai_chat/core/common/mojom/tab_tracker.mojom.h"
+#include "mojo/public/cpp/bindings/self_owned_receiver.h"
+#include "testing/gtest/include/gtest/gtest.h"
+
+namespace ai_chat {
+
+namespace {
+class TabDataObserver : public mojom::TabDataObserver {
+ public:
+  TabDataObserver() = default;
+  ~TabDataObserver() override = default;
+
+  TabDataObserver(const TabDataObserver&) = delete;
+  TabDataObserver& operator=(const TabDataObserver&) = delete;
+
+  TabDataObserver(TabDataObserver&&) = default;
+  TabDataObserver& operator=(TabDataObserver&&) = default;
+
+  // mojom::TabDataObserver
+  void TabDataChanged(std::vector<mojom::TabDataPtr> tabs) override {
+    this->last_tabs_ = std::move(tabs);
+  }
+
+  const std::vector<mojom::TabDataPtr>& last_tabs() const { return last_tabs_; }
+
+ private:
+  std::vector<mojom::TabDataPtr> last_tabs_;
+};
+}  // namespace
+
+class TabTrackerServiceTest : public testing::Test {
+ public:
+  TabTrackerServiceTest() {
+    mojo::PendingRemote<mojom::TabDataObserver> pending_remote;
+    auto observer = std::make_unique<TabDataObserver>();
+    observer_ = observer.get();
+
+    auto receiver = mojo::MakeSelfOwnedReceiver(
+        std::move(observer), pending_remote.InitWithNewPipeAndPassReceiver());
+    service_.AddObserver(std::move(pending_remote));
+  }
+  ~TabTrackerServiceTest() override = default;
+
+  const std::vector<mojom::TabDataPtr>& last_tabs() const {
+    return observer_->last_tabs();
+  }
+
+  void UpdateTab(mojom::TabDataPtr tab) {
+    auto id = tab->id;
+    service_.UpdateTab(id, std::move(tab));
+  }
+
+  void DeleteTab(int id) { service_.UpdateTab(id, nullptr); }
+
+ private:
+  base::test::SingleThreadTaskEnvironment task_environment_;
+
+  TabTrackerService service_;
+  raw_ptr<TabDataObserver> observer_;
+};
+
+TEST_F(TabTrackerServiceTest, TabsCanBeAdded) {
+  auto tab1 = mojom::TabData::New();
+  tab1->title = "One";
+  tab1->url = GURL("https://one.com");
+  tab1->id = 1;
+  UpdateTab(std::move(tab1));
+
+  auto tab2 = mojom::TabData::New();
+  tab2->title = "Two";
+  tab2->url = GURL("https://two.com");
+  tab2->id = 2;
+  UpdateTab(std::move(tab2));
+
+  base::test::RunUntil([&]() { return last_tabs().size() == 2; });
+
+  EXPECT_EQ(last_tabs()[0]->title, "One");
+  EXPECT_EQ(last_tabs()[0]->url, GURL("https://one.com"));
+  EXPECT_EQ(last_tabs()[0]->id, 1);
+
+  EXPECT_EQ(last_tabs()[1]->title, "Two");
+  EXPECT_EQ(last_tabs()[1]->url, GURL("https://two.com"));
+  EXPECT_EQ(last_tabs()[1]->id, 2);
+}
+
+TEST_F(TabTrackerServiceTest, TabsCanBeRemoved) {
+  auto tab1 = mojom::TabData::New();
+  tab1->title = "One";
+  tab1->url = GURL("https://one.com");
+  tab1->id = 1;
+  UpdateTab(std::move(tab1));
+
+  auto tab2 = mojom::TabData::New();
+  tab2->title = "Two";
+  tab2->url = GURL("https://two.com");
+  tab2->id = 2;
+  UpdateTab(std::move(tab2));
+
+  DeleteTab(1);
```

Comment:
what happens if we try to delete a tab that doesn't exist? There should also be a test for removing a tab when there is only one, etc...

Reply:
I see the test for deleting a tab that doesn't exist below.